File: blind-75/graphs/numIslands.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if(grid.empty()) return 0;
        
        int result = 0;
        vector<vector<bool>> visited(grid.size(), vector<bool>(grid[0].size()));
        
        numIslands(0, 0, result, grid, visited);
        return result;
    }

private:
    void numIslands(int i, int j, int count, vector<vector<char>>& grid, vector<vector<bool>>& visited){
        if(i == grid.size() || i < 0 || j == grid[0].size() || j < 0) return;
        
        if(visited[i][j]) return;

        for(auto i = 0; i < grid.size(); i++) {
            for(auto j = 0; j < grid[0].size(); j++) {
                visited[i][j] = true;
                
                if(grid[i][j] - '0'){
                    numIslands(i, j+1, count, grid, visited);
                    numIslands(i+1, j, count, grid, visited);
                    numIslands(i, j-1, count, grid, visited);
                    numIslands(i-1, j, count, grid, visited);

                    ++count;
                }
            }
        }
    }

};

File: blind-75/graphs/num_islands.py

class Solution:

    def __init__(self):
        i = 0

    def numIslands(self, matrix: list[list]) -> int:
        visited = set()

        count = self.__numIslands(matrix, visited, 0, 0)

        return count

    def __numIslands(self, matrix: list[list], visited: set, row: int, column: int):
        oob = (column < 0) or (row < 0) or (column > len(matrix[0])-1) or (row > len(matrix)-1)
        if ((row, column) in visited) or oob:
            return 0
        
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if matrix[row][column] == "1" and (row, column) not in visited:
                    visited.add((row, column))

                    count = self.__numIslands(matrix, visited, row+1, column)
                    count = self.__numIslands(matrix, visited, row-1, column)
                    count = self.__numIslands(matrix, visited, row, column+1)
                    count = self.__numIslands(matrix, visited, row, column-1)

        count += 1

        return count

def main():
    mat = [["1", "0", "1", "0", "1", "0"], ["1", "0", "1", "0", "1", "0"], 
           ["1", "0", "1", "0", "1", "0"]]
    
    s = Solution()

    print(s.numIslands(mat))


if __name__ == "__main__":
    main()
